custom/icds_reports/utils/aggregation_helpers/agg_ls_data.py

```diff
@@ -0,0 +1,214 @@
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+from dateutil.relativedelta import relativedelta
+
+from corehq.apps.userreports.models import StaticDataSourceConfiguration, get_datasource_config
+from corehq.apps.userreports.util import get_table_name
+
+from custom.icds_reports.utils.aggregation_helpers import BaseICDSAggregationHelper, transform_day_to_month
+
+
+class AggLsDataHelper(BaseICDSAggregationHelper):
+
+    base_tablename = 'agg_ls_report'
+    awc_location_ucr = 'static-awc_location'
+    ls_vhnd_ucr = 'static-ls_vhnd_form'
+    ls_home_visit_ucr = 'static-ls_home_visit_forms_filled'
+    ls_awc_mgt_ucr = 'static-awc_mgt_forms'
+
+    def __init__(self, month):
+        self.month_start = transform_day_to_month(month)
+        self.next_month_start = self.month_start + relativedelta(months=1)
+
+    def _tablename_func(self, agg_level):
+        return "{}_{}_{}".format(self.base_tablename, self.month_start.strftime("%Y-%m-%d"), agg_level)
+
+    def drop_table_if_exists(self, agg_level):
+        return """
+        DROP TABLE IF EXISTS "{table_name}"
+        """.format(table_name=self._tablename_func(agg_level))
+
+    def create_child_table(self, agg_level):
+        return """
+        CREATE TABLE "{table_name}" (
+        CHECK (month=DATE %(start_date)s AND aggregation_level={agg_level})
+        ) INHERITS ({base_tablename})
+        """.format(
+            table_name=self._tablename_func(agg_level),
+            base_tablename=self.base_tablename,
+            agg_level=agg_level
+        ), {
+            "start_date": self.month_start
+        }
+
+    @property
+    def tablename(self):
+        return self._tablename_func(4)
+
+    def _ucr_tablename(self, ucr_id):
+        doc_id = StaticDataSourceConfiguration.get_doc_id(self.domain, ucr_id)
+        config, _ = get_datasource_config(doc_id, self.domain)
+        return get_table_name(self.domain, config.table_id)
+
+    def aggregate_query(self):
+        """
+        Returns the base aggregate query which is used to insert all the locations
+        into the LS data table.
+        """
+        return """
```

Comment:
It would be awesome if you could make this all one query with joins instead of inserting dummy data then updating it multiple times. It should be far more performant that way

Reply:
If necessary you can add the ucr aggregation tables like we do for child health and ccs record data to facilitate this.

Reply:
Thought through it. So, making it into a single query using joins seems to make the query unnecessarily complex and long. That too, I don't see it can be done without having 3 selects in a single query because of the following reason and would appreciate help on it:
1.  each of the indicator is coming from different datasource
2. each indicators though has to be grouped by supervisor_id but supervisor_ids can have dis junction between data source like one is present in x datasource but not in y but in z.
3. each indicator have different deciding condition which row to count which not to.

Can we have the scope of keeping this as it is and see if this causes a problem in performance.

Reply:
I think that in this case, independent of performance, consistency is important. It makes the dashboard easier to maintain and to understand. Additionally, even if performance is not a concern at current scale, it will be in the future. This way we will not have to rewrite this query down the line.

As far as the query itself, I think it should be pretty straightforward. You can use the same model we use on other agg tables and create aggregation tables for each ucr, and then do the join with those instead of the ucr itself (like this table https://github.com/dimagi/commcare-hq/blob/master/custom/icds_reports/models/aggregate.py#L1282) and then the agg ccs record